File: src/2744.find-maximum-number-of-string-pairs.2847/solution.ts
function maximumNumberOfStringPairs (words: string[]): number {
    const rwords = words.map(word => word.split('').reverse().join(''));
    const set = new Set(rwords);
    let result = 0;
    for (let i = 0; i < words.length; i++) {
        set.delete(rwords[i]);
        if (set.has(words[i])) {
            result++;
        }
    }
    return result;
}

File: src/2744.find-maximum-number-of-string-pairs.2847/solution.java
class Solution {
    public int maximumNumberOfStringPairs(String[] words) {
        Set<String> set = new HashSet<>();
        int result = 0;
        for (String s : words) {
            char[] chars = s.toCharArray();
            char[] reverse = new char[chars.length];
            for (int i = 0; i < chars.length; i++) {
                reverse[chars.length - i - 1] = chars[i];
            }
            String s2 = String.valueOf(reverse);
            if (set.contains(s2)) {
                result++;
            }
            set.add(s);
        }
        return result;
    }
}